### PR TITLE
use configargparse to process halmos.toml

### PR DESCRIPTION
**Pull request by karmacoma-eth**

### CODE REVIEW

File: src/halmos/sevm.py [hunk 1]
```diff
@@ -1053,6 +1053,7 @@ def empty(cls, addr: BitVecRef, loc: BitVecRef) -> ArrayRef:
     @classmethod
     def init(cls, ex: Exec, addr: Any, loc: BitVecRef) -> None:
         assert_address(addr)
+        print(f"{addr = }")
```
Comment (daejunpark) [inline src/halmos/sevm.py:1056]: ?

**Comment by daejunpark** [comment]
could you explain in which order options are applied?  we have four sources of options:
1. default option values
2. config file
3. halmos custom natspec
4. command line

**Comment by daejunpark** [comment]
what's the meaning of sections? can we have multiple sections in a toml file, and choose which section to use when running halmos? do sections inherit from their parents? what's the naming convention for section inheritance?

File: src/halmos/__main__.py [hunk 2]
```diff
@@ -974,9 +970,8 @@ def run_sequential(run_args: RunArgs) -> List[TestResult]:
             funsig.split("(")[0], funsig, run_args.methodIdentifiers[funsig]
         )
         try:
-            extended_args = extend_args(
-                args, parse_devdoc(funsig, run_args.contract_json)
-            )
+            extended_args = args.extend(parse_devdoc(funsig, run_args.contract_json))
+            debug(extended_args.format_values())
```
[the remark this answers is not in the document]
Comment (karmacoma-eth) [reply src/halmos/__main__.py:974]: nope sorry, some leftover

File: src/halmos/sevm.py (hunk 1 above)
Comment (karmacoma-eth) [reply src/halmos/sevm.py:1056]: another leftover

**Comment by karmacoma-eth** [comment]
> could you explain in which order options are applied? we now have four sources of options:
> 
>     1. default option values
> 
>     2. config file
> 
>     3. halmos custom natspec
> 
>     4. command line
> 
> 
> i think the above order makes sense. wdyt?

almost but IMO it should be:

1. defaults
2. config
3. command line
4. natspec

natspec feels more specific to a contract/function than the command line, which applies to the whole repo

File: halmos.toml [hunk 3]
```diff
@@ -1,9 +1,12 @@
...
 
-[compile-settings]
-array-lengths = 2
\ No newline at end of file
+verbose = 3
+debug = true
+statistics = true
+
+depth = 1_000_000
+storage-layout = 'generic'
```
[the remark this answers is not in the document]
Comment (karmacoma-eth) [reply halmos.toml:12]: yes!

**Comment by karmacoma-eth** [comment]
> what's the semantics of sections? can we have multiple sections in a toml file, and choose which section to use when running halmos? do sections inherit from their parents? if so, what's the naming convention for section inheritance?

left for future work, currently there is a single supported `[global]` section

but going forward I think it would be nice to support profiles, so you could have `[ci]` and `[ci-long]` sections for instance, and then you can enable them via an env var

**Comment by daejunpark** [comment]
> almost but IMO it should be:
> 
> 1. defaults
> 2. config
> 3. command line
> 4. natspec
> 
> natspec feels more specific to a contract/function than the command line, which applies to the whole repo

makes sense. that's also the existing behavior before this pr, right?

actually i was wondering what if users want to quickly override the natspec options using command line, but that might not worth supporting.

**Review by daejunpark — APPROVED** [review]

**Comment by daejunpark** [comment]
tests seem to become flaky. ci failed due to z3 model not found, and that also happened intermittently in my local:
```
z3.z3types.Z3Exception: model is not available
```

**Comment by karmacoma-eth** [comment]
I'm not happy with the hackiness and the flakiness, I'll close this one and put up something cleaner

**Pull request closed without merging**